File: src/main/scala/crusader728/leetcode/pq/KthLargestElementInArray215.scala
package scala.crusader728.leetcode.pq

object KthLargestElementInArray215 {
  def findKthLargest(nums: Array[Int], k: Int): Int = {
    val pq = collection.mutable.PriorityQueue.empty[Int](implicitly[Ordering[Int]].reverse)
    nums.foreach(i => {
      if(pq.size < k) {
        pq.enqueue(i)
      } else {
        if(i <= pq.head) {
          ()
        } else {
          pq.dequeue()
          pq.enqueue(i)
        }
      }
    })

    pq.head
  }
}

File: src/main/withJava/crusader728/leetcode/pq/KthLargestElementInArray215.java
package withJava.crusader728.leetcode.pq;

import java.util.PriorityQueue;

public class KthLargestElementInArray215 {
    public int findKthLargest(int[] nums, int k) {
        if(nums == null || k > nums.length || k < 1) {
            throw new IllegalArgumentException();
        }

        PriorityQueue<Integer> priorityQueue = new PriorityQueue<>();
        for(int n: nums) {
            priorityQueue.offer(n);
            while(priorityQueue.size() > k) {
                priorityQueue.poll();
            }
        }

        return priorityQueue.poll();
    }
}
